actions/actions_whitebox_test.go

```diff
@@ -0,0 +1,184 @@
+package actions
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/fabric8-services/fabric8-wit/gormtestsupport"
+	"github.com/fabric8-services/fabric8-wit/resource"
+	tf "github.com/fabric8-services/fabric8-wit/test/testfixture"
+	"github.com/fabric8-services/fabric8-wit/workitem"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+func TestSuiteAction(t *testing.T) {
+	resource.Require(t, resource.Database)
+	suite.Run(t, &ActionSuite{DBTestSuite: gormtestsupport.NewDBTestSuite()})
+}
+
+type ActionSuite struct {
+	gormtestsupport.DBTestSuite
+}
+
+func createWICopy(wi workitem.WorkItem, state string, boardcolumns []interface{}) workitem.WorkItem {
+	var wiCopy workitem.WorkItem
+	wiCopy.ID = wi.ID
+	wiCopy.SpaceID = wi.SpaceID
+	wiCopy.Type = wi.Type
+	wiCopy.Number = wi.Number
+	wiCopy.Fields = map[string]interface{}{}
+	for k := range wi.Fields {
+		wiCopy.Fields[k] = wi.Fields[k]
+	}
+	wiCopy.Fields[workitem.SystemState] = state
+	wiCopy.Fields[workitem.SystemBoardcolumns] = boardcolumns
+	return wiCopy
+}
+
+func (s *ActionSuite) TestChangeSet() {
+	fxt := tf.NewTestFixture(s.T(), s.DB, tf.CreateWorkItemEnvironment(), tf.WorkItems(2))
+
+	s.T().Run("different ID", func(t *testing.T) {
+		_, err := fxt.WorkItems[0].ChangeSet(*fxt.WorkItems[1])
+		require.Error(t, err)
+	})
+
+	s.T().Run("same instance", func(t *testing.T) {
+		changes, err := fxt.WorkItems[0].ChangeSet(*fxt.WorkItems[0])
+		require.NoError(t, err)
+		require.Empty(t, changes)
+	})
+
+	s.T().Run("no changes, same column order", func(t *testing.T) {
+		wiCopy := createWICopy(*fxt.WorkItems[0], "new", []interface{}{"bcid0", "bcid1"})
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{"bcid0", "bcid1"}
+		changes, err := fxt.WorkItems[0].ChangeSet(wiCopy)
+		require.NoError(t, err)
+		require.Empty(t, changes)
+	})
+
+	s.T().Run("no changes, mixed column order", func(t *testing.T) {
+		wiCopy := createWICopy(*fxt.WorkItems[0], "new", []interface{}{"bcid1", "bcid0"})
```

Comment:
Same as above. At that level, these are strings, not UUIDs, and the code expects strings. This is supposed to be generic.

Reply:
No it is not supposed to be generic. You're misunderstanding this. I want you to put strings with UUIDs inside here instead of just arbitrary strings. `uuid.NewV4().String()` is your friend. We will check internally if a board column is a UUID. So this code from the `workitem/simple_type.go` will soon be gone:

```go
case KindString, KindUser, KindIteration, KindArea, KindLabel, KindBoardColumn:
```

Instead we will treat strings as strings and for the rest we will accept a UUID or a string with a UUID only. That's the reason I ask you to please replace this.

Reply:
So you're saying that the code will soon only accept string-stored UUIDs that follow RFC 4122? That seems *very* weird to me. What value does this have?